mindmeld/models/containers.py

```diff
@@ -0,0 +1,462 @@
+        return word_to_embedding
+
+    def _extract_embeddings(self):
+        if self.model_id not in GloVeEmbeddingsContainer.CONTAINER_LOOKUP:
+
+            file_location = self.token_pretrained_embedding_filepath
+
+            if file_location and os.path.isfile(file_location):
+                msg = f"Extracting embeddings from provided file location {str(file_location)}."
+                logger.info(msg)
+                with open(file_location, "r") as embedding_file:
+                    word_to_embedding = self._extract_and_map(embedding_file)
+            else:
+                if file_location:
+                    logger.info("Provided file location %s does not exist.", str(file_location))
+                file_name = GloVeEmbeddingsContainer.EMBEDDING_FILE_PATH_TEMPLATE.format(
+                    self.token_dimension)
+                if os.path.isfile(EMBEDDINGS_FILE_PATH):
+                    msg = f"Extracting embeddings from default folder location " \
+                          f"{EMBEDDINGS_FILE_PATH}."
+                    logger.info(msg)
+                    try:
+                        zip_file_object = zipfile.ZipFile(EMBEDDINGS_FILE_PATH, "r")
+                        with zip_file_object.open(file_name) as embedding_file:
+                            word_to_embedding = self._extract_and_map(embedding_file)
+                    except zipfile.BadZipFile:
+                        logger.warning(
+                            "%s is corrupt. Deleting the zip file and attempting to"
+                            " download the embedding file again",
+                            EMBEDDINGS_FILE_PATH,
+                        )
+                        os.remove(EMBEDDINGS_FILE_PATH)
+                        self._extract_embeddings()
+                    except IOError as e:
+                        logger.error(
+                            "An error occurred when reading %s zip file. The file might"
+                            " be corrupt, so try deleting the file and running the program "
+                            "again",
+                            EMBEDDINGS_FILE_PATH,
+                        )
+                        raise IOError("Failed to load embeddings.") from e
+                else:
+                    logger.info("Default folder location %s does not exist.", EMBEDDINGS_FILE_PATH)
+                    zip_file_object = self._download_embeddings_and_return_zip_handle()
+                    if not zip_file_object:
+                        raise EmbeddingDownloadError("Failed to download embeddings.")
+                    with zip_file_object.open(file_name) as embedding_file:
+                        word_to_embedding = self._extract_and_map(embedding_file)
+
+            GloVeEmbeddingsContainer.CONTAINER_LOOKUP[self.model_id] = word_to_embedding
+
+        return GloVeEmbeddingsContainer.CONTAINER_LOOKUP[self.model_id]
+
+
+class SentenceTransformersContainer:
```

Comment:
Do you think we can add some consistency to the containers file? Like an abstract class with the extraction, get_model, encoder functionalities etc. set up in a default class and then inherited by these different model classes? Might be good to have that base structure in place for adding containers or modifying overall structure across the set of containers at a later point.

Reply:
Yes, we can definitely do that. I worked on it briefly sometime back but kept it as low priority for now because wherever we are using these containers in our codebase, we mostly don't use them interchangeably, meaning we might not need a factory like class structure. Hence I made sure that at least the skeleton of all containers look similar so even if we plan to refactor like you mentioned, it should be mostly staright-forward.